### PR TITLE
Fix heap-buffer overflow from FPRDecodeString 

### LINKED ISSUE #8628: crash on FPRDecodeString when using Thread Sanitizer
### Step 0: Are you in the right place?

  * For issues or feature requests related to __the code in this repository__
    file a Github issue.
    * If this is a __feature request__ please use the Feature Request template.
  * For general technical questions, post a question on [StackOverflow](http://stackoverflow.com/)
    with the `firebase` tag.
  * For general (non-iOS) Firebase discussion, use the [firebase-talk](https://groups.google.com/forum/#!forum/firebase-talk)
    google group.
  * For backend issues, console issues, and other non-SDK help that does not fall under one
    of the above categories, reach out to
    [Firebase Support](https://firebase.google.com/support/).
  * Once you've read this section and determined that your issue is appropriate for
    this repository, please delete this section.

### [REQUIRED] Step 1: Describe your environment

  * Xcode version: 12.5.1
  * Firebase SDK version: 8.6.0
  * Installation method: `CocoaPods` (select one)
  * Firebase Component: Analytics, Crashlytics, Performance

#### Steps to reproduce:

What happened? How can we make the problem occur?
This could be a description, log/console output, etc.

For reproducing, I just check the Adress Sanitizer in Edit Scheme > Run and I run my project, it crashes systematically.

Capture d’écran 2021-09-07 à 14 05 10
Capture d’écran 2021-09-07 à 14 05 24

### PR DESCRIPTION
**Pull request by jeremyjiang-dev**
This is to fix #8628 and #8577 
The issue was not reproducible.
The fix is to remove the usage of FPRDecodeString  in the code.

### CODE REVIEW

**Review by visumickey — COMMENTED** [review]
@jeremyjiang-dev Since FPRClient is the only place where we are using FPRDecodeString, we might be safe here. But, I think we need to get to the root cause if the code exists in our ecosystem. As an alternate, do you think we can FPRDecodeString as an API? This is not used anywhere else except for unit tests.

Heap overflow occurs when there is a byte array conversion to NSData. If the byte array does not contain a 'null' or '\0' at the end, there is no way for the system to undertsand where the character array ends.

One way to attempt to reproduce this issue would be to force a really long character array to see if FPRDecodeString cannot decode it.

**Comment by jeremyjiang-dev** [comment]
> @jeremyjiang-dev Since FPRClient is the only place where we are using FPRDecodeString, we might be safe here. But, I think we need to get to the root cause if the code exists in our ecosystem. As an alternate, do you think we can FPRDecodeString as an API? This is not used anywhere else except for unit tests.
> 
> Heap overflow occurs when there is a byte array conversion to NSData. If the byte array does not contain a 'null' or '\0' at the end, there is no way for the system to undertsand where the character array ends.
> 
> One way to attempt to reproduce this issue would be to force a really long character array to see if FPRDecodeString cannot decode it.

Discussed offline. I moved the decoding methods to testUtil and note the methods may cause heap-buffer overflow.

File: FirebasePerformance/CHANGELOG.md [hunk 1]
```diff
@@ -1,6 +1,7 @@
 # Pending
 * Create a random number of delay for remote config fetch during app starts.
 * Fix log spamming when Firebase Performance is disabled. (#8423)
+* Fix heap-heap-buffer overflow when decoding strings. (#8628)
```
Comment (visumickey) [inline FirebasePerformance/CHANGELOG.md:4]: I thought there were 2 bugs related to this. Can you tag both of them here?

File: FirebasePerformance/Tests/Unit/FPRTestUtils.m [hunk 2]
```diff
@@ -153,4 +154,39 @@ + (GDTCOREvent *)createRandomNetworkGDTEvent:(NSString *)url {
   return gdtEvent;
 }
 
+#pragma mark - Decode nanoPb pbData
+NSData *FPRDecodeData(pb_bytes_array_t *pbData) {
```
Comment (visumickey) [inline FirebasePerformance/Tests/Unit/FPRTestUtils.m:158]: nit: New line here.

**Review by visumickey — APPROVED** [review]

**Review by leotianlizhan — APPROVED** [review]
LGTM

File: FirebasePerformance/CHANGELOG.md (hunk 1 above)
Comment (jeremyjiang-dev) [reply FirebasePerformance/CHANGELOG.md:4]: Done.

File: FirebasePerformance/Tests/Unit/FPRTestUtils.m (hunk 2 above)
Comment (jeremyjiang-dev) [reply FirebasePerformance/Tests/Unit/FPRTestUtils.m:159]: Done

**Pull request merged**